### PR TITLE
Stop using data classes in public API

### PR DESCRIPTION
**Pull request by 3flex**
https://kotlinlang.org/docs/jvm-api-guidelines-backward-compatibility.html#don-t-use-data-classes-in-an-api

> it's better not to use data classes in your API because almost any change in them breaks source, binary, or behavioral compatibility.

Instead I've introduced the Poko plugin where required because calling code used equality checks. This plugin generates overrides for `equals`,  `hashCode` and `toString` but does not generate the `componentX` or `copy` functions that lead to backwards compatibility issues.

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt [hunk 1]
```diff
@@ -11,8 +11,8 @@ import org.jetbrains.kotlin.psi.psiUtil.getNonStrictParentOfType
 /**
  * Stores information about a specific code fragment.
  */
-data class Entity(
-    private val name: String,
+class Entity(
+    val name: String,
```
Comment (BraisGabin) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt:15]: Is this change intended? Shouldn't it remain as private?

**Comment by BraisGabin** [comment]
<3 good one! I had this on my todo list.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt (hunk 1 above)
Comment (3flex) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt:15]: It's deliberate - to copy to a new instance with changed values this needs to be readable so the value can be copied to the new instance.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Debt.kt [hunk 2]
```diff
@@ -4,7 +4,7 @@ package io.gitlab.arturbosch.detekt.api
  * Debt describes the estimated amount of work needed to fix a given issue.
  */
 @Suppress("MagicNumber")
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Debt.kt:6]: Btw, is this `Suppress` statement still needed?

File: detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/CorrectableRulesFirstSpec.kt [hunk 3]
```diff
@@ -19,14 +19,14 @@ class CorrectableRulesFirstSpec {
         var actualLastRuleId = ""
 
         class First(config: Config) : Rule(config) {
-            override val issue: Issue = justAnIssue.copy(id = "NonCorrectable")
+            override val issue: Issue = Issue("NonCorrectable", justAnIssue.description, justAnIssue.debt)
```
Comment (schalkms) [inline detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/CorrectableRulesFirstSpec.kt:22]:
```suggestion
            override val issue: Issue = Issue("NonCorrectable", "", Debt.FIVE_MINS)
```

and remove the variable `justAnIssue`

File: detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/CorrectableRulesFirstSpec.kt [hunk 4]
```diff
@@ -19,14 +19,14 @@ class CorrectableRulesFirstSpec {
...
             override fun visitClass(klass: KtClass) {
                 actualLastRuleId = issue.id
             }
         }
 
         class Last(config: Config) : Rule(config) {
-            override val issue: Issue = justAnIssue.copy(id = "Correctable")
+            override val issue: Issue = Issue("Correctable", justAnIssue.description, justAnIssue.debt)
```
Comment (schalkms) [inline detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/CorrectableRulesFirstSpec.kt:29]:
```suggestion
            override val issue: Issue = Issue("Correctable", "", Debt.FIVE_MINS)
```

**Review by schalkms — APPROVED** [review]
👍

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt (hunk 1 above)
Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt:15]: But do we need to copy them?

Comment (3flex) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt:15]: I just did a simple refactoring to get rid of the data classes without doing anything more at the call sites. It's possible that copying isn't required. I can take a closer look and see if we can keep this private if that's a concern?

Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt:15]: I think that this one should be even removed (on later refactors) so it will help to keep it private. But if it gives too much work, don't worry about it.

**Review by BraisGabin — APPROVED** [review]
